addons/mail/controllers/discuss.py

```diff
@@ -573,3 +573,41 @@ def mail_guest_update_name(self, guest_id, name):
         if guest_to_rename_sudo != guest and not request.env.user._is_admin():
             raise NotFound()
         guest_to_rename_sudo._update_name(name)
+
+    # --------------------------------------------------------------------------
+    # Link preview API
+    # --------------------------------------------------------------------------
+
+    @http.route('/mail/link_preview', type='json', auth='public')
+    def link_preview(self, url, channel_id):
+        channel_partner = request.env['mail.channel.partner']
+        if request.env.user._is_public():
+            channel_partner = request.env['mail.channel.partner']._get_as_sudo_from_request_or_raise(request=request, channel_id=int(channel_id))
+        link_preview_data = request.env['ir.attachment']._get_link_preview_from_url(url)
+        if link_preview_data:
+            link_preview_data['res_model'] = 'mail.compose.message'
+            link_preview_data['res_id'] = 0
+
+        if channel_partner.env.user.share:
+            # Only generate the access token if absolutely necessary (= not for internal user).
+            link_preview_data['access_token'] = channel_partner.env['ir.attachment']._generate_access_token()
+
+        if link_preview_data:
+            attachment = request.env['ir.attachment'].sudo().create(link_preview_data)
+            attachment_formated = attachment._attachment_format(commands=True)[0]
+            if link_preview_data.get('access_token'):
+                attachment_formated['accessToken'] = link_preview_data['access_token']
+            return attachment_formated
```

Comment:
move to method "create link preview ..." rather than kept here?
Try to keep controller only for authenticate, and have business code in models.